swarm/src/lib.rs

```diff
@@ -497,12 +503,54 @@ impl<TBehaviour> Swarm<TBehaviour>
             addresses_from_opts
         };
 
-        let transport_dial_opts = self.convert_dial_opts(&dial_opts);
+        // allows to map a protocol, if the tag is one, to an idx. Allows for O(1) operations on presence checking.
+        // This might be a premature optimization and is prone to errors. If Protocol::tag would be const one could write a simpler variant, with comparable performance.
+        fn protocol_idx(tag: &str) -> Option<usize> {
+            match tag {
+                "memory" => Some(0),
+                "onion" => Some(1),
+                "onion3" => Some(2),
+                "quic" => Some(3),
+                "quic-v1" => Some(4),
+                "tcp" => Some(5),
+                "udp" => Some(6),
+                "webtransport" => Some(7),
+                "ws" => Some(8),
+                "wss" => Some(9),
+                _ => None,
+            }
+        }
```

Comment:
Here is what I would suggest:

- Every time we get a listen address, we compute its protocol stack, trim unnecessary things from it like `/ip4` etc and then store it in `HashMap` (the value is a counter of how many listen addresses of this type we've got).
- When we want to dial, we apply the same treatment to the address, lookup the value in the hashmap and check if it is greater `0`.

I'd suggest we add some good logging and several unit tests to make sure we are trimming the addresses correctly.

Reply:
I think I understand now. That makes much more sense. But why a HashMap? Why not just a hash set? Since the question is: Is there a listener for this protocol stack present or not? And not how many listeners are there for a given protocol stack. 

Should we compute this on premise, every time we dial (which reduces complexity) or every time listeners change?

Reply:
Listen addresses can expire, see `TransportEvent::AddressExpired`. If we don't use a `HashMap` and count the addresses per protocol stack, we would end up assuming that we don't have a listen address if we previously got 2 `TransportEvent::NewAddress` and 1 `TransportEvent::AddressExpired`.

> Should we compute this on premise, every time we dial (which reduces complexity) or every time listeners change?

Intuitively, I would have said compute it when listen addresses change because that happens a lot less compared to dialing. It is a trade-off of memory vs CPU power and I'd probably favor using more memory.

Reply:
> Listen addresses can expire, see TransportEvent::AddressExpired. If we don't use a HashMap and count the addresses per protocol stack, we would end up assuming that we don't have a listen address if we previously got 2 TransportEvent::NewAddress and 1 TransportEvent::AddressExpired.

I understand the problem, however that just happens if we calculate it every time listeners change. Just to test out I wrote an implementation with HashSet and on premise calculation.

> It is a trade-off of memory vs CPU power and I'd probably favor using more memory.

I'm very open to keeping it in memory, but I want to point out that it will require a more complex logic (since we have to track how many listen addresses are present) and on the other hand I don't see a big problem performance wise to calculate the HashSet when required. The data structure is very small and probably benefits from good locality, thus it will likely have a very small performance penalty.

However I don't like both HashMap and HashSet. Since the keyspace is basically known and there are likely very few entries. How do you feel about using a RadixTree instead?

Reply:
> > Listen addresses can expire, see TransportEvent::AddressExpired. If we don't use a HashMap and count the addresses per protocol stack, we would end up assuming that we don't have a listen address if we previously got 2 TransportEvent::NewAddress and 1 TransportEvent::AddressExpired.
> 
> I understand the problem, however that just happens if we calculate it every time listeners change. Just to test out I wrote an implementation with HashSet and on premise calculation.

Fair, good call!

> > It is a trade-off of memory vs CPU power and I'd probably favor using more memory.
> 
> I'm very open to keeping it in memory, but I want to point out that it will require a more complex logic (since we have to track how many listen addresses are present) and on the other hand I don't see a big problem performance wise to calculate the HashSet when required. The data structure is very small and probably benefits from good locality, thus it will likely have a very small performance penalty.
> 
> However I don't like both HashMap and HashSet. Since the keyspace is basically known and there are likely very few entries. How do you feel about using a RadixTree or BTree instead?
> 
> I think I will write a version using constant updates.

Hmm, I think I'd worry about code simplicity first before going down optimising the data structure even more, esp. without benchmarks.

I see your point in regards to the added complexity of state. I guess it would really be simpler if we did it inline.
Maybe something like:

```rust
let is_listening_on_same_transport = self.listen_addresses.iter().map(...).any(...);
```

Where we map to a pruned protocol stack and compare them in `any` against the to be dialed address?

Reply:
That sounds good. If we see a big performance problem there the optimization is very easy. I adjusted the implementation in the PR and looking forward to your review.